jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala

```diff
@@ -331,34 +333,66 @@ object XGBoost extends Serializable {
-      val trackerReturnVal = parallelismTracker.execute(tracker.waitFor(0L))
-      logger.info(s"Rabit returns with exit code $trackerReturnVal")
-      val model = postTrackerReturnProcessing(trackerReturnVal, boostersAndMetrics,
-        sparkJobThread, isClsTask)
-      if (isClsTask){
-        model.asInstanceOf[XGBoostClassificationModel].numOfClasses =
-          params.getOrElse("num_class", "2").toString.toInt
+    val hdfsTmpPath: String = params.get("hdfs_tmp_path") match {
+      case None => null
+      case Some(path: String) => if (path.nonEmpty) {
+        path
+      } else {
+        null
       }
-      model
-    } finally {
-      tracker.stop()
+      case _ => throw new IllegalArgumentException("parameter \"hdfs_tmp_path\" must be" +
+        " an instance of String.")
+    }
+
+    val savingFreq: Int = params.get("saving_frequency") match {
+      case None => 0
+      case Some(freq: Int) => freq
+      case _ => throw new IllegalArgumentException("parameter \"saving_frequency\" must be" +
+        " an instance of Int.")
     }
+
+    val savingRounds: Seq[Int] = if (hdfsTmpPath == null || savingFreq <= 0){
+      Seq(round)
+    } else {
+      (savingFreq until round by savingFreq) :+ round
+    }
+    savingRounds.map {
+      savingRound: Int =>
+        val tracker = startTracker(nWorkers, trackerConf)
+        try {
+          val sc = trainingData.sparkContext
+          val parallelismTracker = new SparkParallelismTracker(sc, timeoutRequestWorkers, nWorkers)
+          val overriddenParams = overrideParamsAccordingToTaskCPUs(params, sc)
+          val boostersAndMetrics = buildDistributedBoosters(trainingData, overriddenParams,
+            tracker.getWorkerEnvs, nWorkers, savingRound, obj, eval, useExternalMemory, missing,
+            hdfsTmpPath)
+          val sparkJobThread = new Thread() {
+            override def run() {
+              // force the job
+              boostersAndMetrics.foreachPartition(() => _)
+            }
+          }
+          sparkJobThread.setUncaughtExceptionHandler(tracker)
+          sparkJobThread.start()
+          val isClsTask = isClassificationTask(params)
+          val trackerReturnVal = parallelismTracker.execute(tracker.waitFor(0L))
+          logger.info(s"Rabit returns with exit code $trackerReturnVal")
+          val model = postTrackerReturnProcessing(trackerReturnVal, boostersAndMetrics,
+            sparkJobThread, isClsTask)
+          if (isClsTask){
+            model.asInstanceOf[XGBoostClassificationModel].numOfClasses =
+              params.getOrElse("num_class", "2").toString.toInt
+          }
+          if (hdfsTmpPath != null && savingRound < round) {
+            val fullPath = s"$hdfsTmpPath/${model.version}$modelSuffix"
+            logger.info(s"Saving temporary model to $fullPath")
+            model.saveModelAsHadoopFile(fullPath)(sc)
+          }
+          model
+      } finally {
+        tracker.stop()
+      }
+    }.last
```

Comment:
we may want to clean up all temporary files when we finish the last round training,

Reply:
Cleaned during temp model saving.